utils/das_data_sorter.py

```diff
@@ -128,14 +127,97 @@ def gps_data(self, data:pd.Series) -> pd.Series:
                 new_gps_data.append(0)
-        
+
         return new_gps_data
     
+    def mean_smooth_caller(self, n:int) -> None:
+        '''Calls __mean_smooth function to smooth the data in instance variables, without requiring to refer to the variables themselves.'''
+        if n < 3 or n > len(data):
+            raise ValueError("Number of smoothing points must be at least 3 and less than the length of the data set to perform smoothing.")
+        
+        # passes instance variables as an argument in smoothing function
+        for variable in self.__dict__:
+            if variable != "indexes": # excludes self.indexes
+                self.__dict__[variable] = self.__mean_smooth(self.__dict__[variable], n)
+
+    def __mean_smooth(self, data:pd.Series, n:int) -> pd.Series:
+        '''Returns an array of smoothed data based on number of data points taken to smooth. 
+        
+        For an odd number N, the data points are simply averaged. 
+        Whereas for an even number N, the data points are averaged, then centered. This is because the data point will misalign with integer
+        numbers of time if not done.'''
+        smooth_data_array = [] 
+        if n % 2 != 0: # Mean Smoothing for odd number N
+            for i in range(len(data) - n + 1):
+                data_points = data[i:i+n]
+                new_data_point = self.mean(data_points)
+                new_data_point = round(new_data_point, ndigits=2)
+                smooth_data_array.append(new_data_point)
+        else: # Mean Smoothing for even number N
+            temp_array = []
+            for i in range(len(data) - n + 1):
+                data_points = data[i:i+N]
+                new_data_point = self.mean(data_points)
+                temp_array.append(new_data_point)
+            
+            # an extra step to centre the data by averaging adjacent data points
+            for j in range(len(temp_array) - 1): 
+                first = temp_array[j]
+                second = temp_array[j+1]
+                new_data_point = self.mean([first,second])
+                new_data_point = round(new_data_point, ndigits=2)
+                smooth_data_array.append(new_data_point)
+        
+        return smooth_data_array 
+    
+    def median_smooth_caller(self, n:int) -> None:
+        '''Calls __median_smooth function to smooth the data in instance variables, without requiring to refer to the variables themselves.'''
+        if n < 3 or n > len(data):
+            raise ValueError("Number of smoothing points must be at least 3 and less than the length of the data set to perform smoothing.")
+        
+        # passes instance variables as an argument in smoothing function
+        for variable in self.__dict__:
+            if variable != "indexes": # excludes self.indexes
+                self.__dict__[variable] = self.__median_smooth(self.__dict__[variable], n)
+    
+    def __median_smooth(self, data:pd.Series, n:int) -> pd.Series:
```

Comment:
Looking back at my older comments I realise it was me that told them to split them up... Sorry about that! I guess now with them split up it became easier to see how similar the code was and how it could be combined but more concisely with less repetition than before. That's part of programming I guess!

Reply:
I think I might leave the two `__mean_smooth` and `__median_smooth` functions separate, just for readability and its easier to maintain. However, I could do these for the 'caller functions'. Would this ruin the readability in `if __name__ == '__main__'`?

Reply:
```python
def smooth_caller(self, data:pd.Series, n:int, average_func) -> pd.Series:
    if average_func == "mean":
        self.__mean_smooth(data, n)
    elif average_func == "median":
        self.__median_smooth(data, n)
    else:
        raise ValueError
...
if __name__ == '__main__':
    if args.smooth:
        smooth_caller(args.data, args.n, args.smooth)
```

So this will keep the two smoothing functions separate, but the caller function is the same. I know you've said to get rid of the 'caller' part, I just used it as an example for now. Thoughts?

Reply:
While I understand what you're saying about increased readability, I don't think it'd be easier to maintain. Interested to know why you disagree!

As I see it, we've got two almost identical functions. If we need to change how one of them works in the future, we'll almost certainly want to also change the other one. This means we'd have to remember to change code in two locations for the one change, which makes it harder to maintain.

Reply:
Oh right, it is harder to maintain. I was just giving a word vomit of word thatll make me sound like I know what im doing lol, don't mind me

Reply:
Gonna have to agree with @hallgchris and say that I would want to have one single function that takes in the smoothing function as a parameter to this function. This is because the code is actually duplicated and like stated before, will be harder to maintain. This should also resolve the issue below of applying median vs mean to a two element array because we just apply  the stated smoothing function.